planner/cascades/transformation_rules.go

```diff
@@ -634,3 +635,50 @@ func (r *EliminateProjection) OnTransform(old *memo.ExprIter) (newExprs []*memo.
 	}
 	return finalGroupExprs, true, false, nil
 }
+
+// MergeAdjacentProjection merge the adjacent projection.
+type MergeAdjacentProjection struct {
+	baseRule
+}
+
+// NewRuleMergeAdjacentProjection creates a new Transformation MergeAdjacentProjection.
+// The pattern of this rule is `Projection -> Projection`.
+func NewRuleMergeAdjacentProjection() Transformation {
+	rule := &MergeAdjacentProjection{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandProjection,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandProjection, memo.EngineTiDBOnly),
```

Comment:
Here we may not check the child pattern is `OperandProjection`. Because for some situations, we may inject projection for other patterns. Such as `IndexLookUpReader`, if it requires keeping order and the primary key is not handle, the `IndexLookUpReader` will be added one projection.

Reply:
But `IndexLookUpReader` adds that `Projection` in the `ImplementationPhase`, I think. This rule only works in the `TransformationPhase`?

Reply:
Yeah, you are right. It seems after implementation we also need do once projection elimination like the old planner.

Reply:
Yes, we might need that again.

Reply:
That's the weak point for the current cascades planner framework. We still separate the implementation rules and transformation rules.